File: 2024/Day_1/day_1.py


import os 
from collections import Counter

with open(os.getcwd() + '/2024/Day_1/data.txt', 'r') as file:
    lines = file.readlines()
    formatted = [line.split('\n') for line in lines]
    left_distance = []
    right_distance = []
    for line in formatted:
        first, second = line[0].split('   ')
        left_distance.append(int(first))
        right_distance.append(int(second))
    left_distance.sort()
    right_distance.sort()

def part_one():
    total = 0
    for i in range(len(left_distance)):
        total += abs(left_distance[i] - right_distance[i])
    return f"Total difference for Part 1 is {total}"


def part_two():
    similarity_score = 0
    right_count = Counter(right_distance)
    for n in left_distance:
        if n in right_count:
            similarity_score += right_count[n] * n
    return f"Similarity score for Part 2 is {similarity_score}"


print(part_one())
print(part_two())

File: 2024/Day_1/day_1.ts
import * as fs from 'fs';
import * as path from 'path';

interface Counter {
    [key: number]: number;
}

function readData(): [number[], number[]] {
    const filePath = path.join(process.cwd(), '2024', 'Day_1', 'data.txt');
    const fileContent = fs.readFileSync(filePath, 'utf-8');
    const lines = fileContent.split('\n');

    const leftDistance: number[] = [];
    const rightDistance: number[] = [];

    lines.forEach(line => {
        if (line.trim()) {
            const [first, second] = line.split('   ').map(s => parseInt(s.trim()));
            if (!isNaN(first) && !isNaN(second)) {
                leftDistance.push(first);
                rightDistance.push(second);
            }
        }
    });

    return [
        leftDistance.sort((a, b) => a - b),
        rightDistance.sort((a, b) => a - b)
    ];
}

function partOne(leftDistance: number[], rightDistance: number[]): string {
    const total = leftDistance.reduce((sum, curr, i) => {
        return sum + Math.abs(curr - rightDistance[i]);
    }, 0);

    return `Total difference for Part 1 is ${total}`;
}

function partTwo(leftDistance: number[], rightDistance: number[]): string {
    let similarityScore = 0;
    const rightCount: Counter = rightDistance.reduce((acc, curr) => {
        acc[curr] = (acc[curr] || 0) + 1;
        return acc;
    }, {} as Counter);

    leftDistance.forEach(n => {
        if (rightCount[n]) {
            similarityScore += rightCount[n] * n;
        }
    });

    return `Similarity score for Part 2 is ${similarityScore}`;
}

try {
    const [leftDistance, rightDistance] = readData();
    console.log(partOne(leftDistance, rightDistance));
    console.log(partTwo(leftDistance, rightDistance));
} catch (error) {
    console.error('Error:', error);
}